internal/s3client/blobclient.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package s3client
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/juju/errors"
+)
+
+// Session is the interface that wraps the GetObject method.
+type Session interface {
+	// GetObject returns a reader for the specified object.
+	GetObject(ctx context.Context, bucketName, objectName string) (io.ReadCloser, int64, string, error)
+}
+
+// Blobs is the client desienged to interact with the
+// s3 compatible object store hosted by the apiserver
+type Blobs struct {
+	session Session
+}
+
+// NewBlobsS3Client creates a client to interact with charm blobs stored
+// on the apiserver's s3 compatible object store.
+func NewBlobsS3Client(session Session) *Blobs {
+	return &Blobs{session: session}
+}
+
+// GetCharm retrieves a charm from the S3-compatible object store hosted
+// by the apiserver. Returns an archived charm as a stream of bytes
+func (c *Blobs) GetCharm(ctx context.Context, modelUUID, charmRef string) (io.ReadCloser, error) {
+	bucketName := fmt.Sprintf("model-%s", modelUUID)
+	objectName := fmt.Sprintf("charms/%s", charmRef)
```

Comment:
This was probably discussed/addressed in previous PRs, but why do we have to treat the charms as a specific object? can't we have `{bucket}/{sha256}` and let the caller know what to expect?

Reply:
The `charms` endpoint ensures that the charm exists and is associated with an application. The objects handler has no concerns. They're for different reasons, although the charms handler could be generalised in the future.